lessons/r/dplyrmagrittr/lesson.Rmd

```diff
@@ -0,0 +1,351 @@
+
+The overall philosophy binding `dplyr` together is rather simple - save for a select few exceptions, nearly every function in the package follows the convention of taking in a data frame as input and returning a modified data frame as output. Coupled with the pipe operator from the `magrittr` package, chaining `dplyr` functions together makes data frame manipulation an absolute breeze.
+
+## Meet the pipe
+
+The R pipe, or `%>%` (Ctrl/Cmd + Shift + M in RStudio) initially began life outside of `dplyr`, finding its R beginnings in the `magrittr` package instead. Its creator, Stefan Milton Bache, was inspired by the pipe operator in F# (`|>`) in addition to its Unix shell equivalent (`|`).
+
+Much like in those examples, the idea behind the `magrittr` pipe is also relatively simple: it essentially stands for 'evaluate the left hand side, and feed the result as input to the function on the right hand side'. For functions that take multiple arguments, the pipe will feed it to the first one by default. Like so:
+
+```{r}
+library(magrittr)
+
+x <- c(1,2)
+
+sum(x, 3)
+x %>% sum(3)
+sum(x, 3) == x %>% sum(3)
+```
+
+Of course, sometimes we want our LHS to be fed into a different argument that just so happens to not be the first one. In these cases, we use `.` to tell R where to send the LHS instead:
+
+```{r}
+
+seq(1, 10, 2)
+
+1 %>% seq(10, 2)
+
+10 %>% seq(1, ., 2)
+
+```
+
+Given the `dplyr` concept of each function taking in a data frame and returning a modified version, it made a lot of sense to integrate the pipe into the `dplyr` workflow. This way, a given data frame would be 'piped' through a series of functions one after the other in order to obtain a specific desired output.
+
+To summarise:
+
+```{r, eval = FALSE}
+
+f(x) == x %>% f()
+
+f(x, y) == x %>% f(y)
+
+f(x, y) == y %>% f(x, .)
+
+```
+
+
+## Tidy data
+
+One final thing to note before we delve into `dplyr` itself is the concept of _tidy data_. This basically describes a data set where each column is a variable or feature of the data, and every row is a single observation. Fisher and Anderson's `iris`, which we'll be using in today's lesson, is a good example of a tidy dataset.
+
+
+```{r}
+head(iris)
+```
+
+The `iris` dataset gives measurements of sepal length/width and petal length/width for 50 flower samples from each of _Iris setosa_, _Iris versicolor_, and _Iris virginica_. Notice how each row corresponds to measurements from a single flower sample, and each column represents a specific feature of that flower. 
+
+`dplyr` is optimized for use with tidy data. However, real world data are rarely in a tidy format from the get-go. For this reason, Hadley and co. have also developed the `tidyr` package, which provides useful functions for reshaping data. We won't be covering `tidyr` in today's lesson, but a wealth of tutorials are available online.
```

Comment:
Technically, dplyr isn't necessarily optimized for tidy datasets. You use dplyr and tidyr together to get a messy dataset into a tidy dataset. Not really that important though.